### PR TITLE
ENH: stats: override _rvs for nhypergeom

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-10106

#### What does this implement/fix?
Overrides `_rvs` for `nhypergeom`, which was the slowest among the discrete distributions.

The approach is to create a fast inverse cdf with `interp1d` then do inverse transform sampling. The support of the distribution is the integers `0, 1, ..., n)` so this tends to be quite fast, and it's exactly as accurate as the `cdf`.

My guess is that `ppf` is more often called for one or a few points whereas it is common for many random variates to be generated at once, so I put the cdf inversion inside `_rvs` only rather than in `_ppf`.

This could be further improved, of course (e.g. use the default if `c * np.prod(size) < n`, where `c` is the expected number of `cdf` evaluations needed for the default `ppf` method to converge), but I'm not sure if that's important.

#### Additional information
~~Mostly I did this to better understand how `rvs` works with `ndarray` shape parameters. If this approach is OK, I will try to break the approach into a function so that there is an easy, standard way of getting the desired broadcasting behavior from an _rvs function that natively handles only scalar shape parameters.~~

_Update: now uses the decorator written for the `nchypergeom_...`s to do the broadcasting._

### CODE REVIEW

File: scipy/stats/_discrete_distns.py [hunk 1]
```diff
@@ -653,6 +654,36 @@ def _argcheck(self, M, n, r):
...
+            return rvs
+
+        M, n, r = np.broadcast_arrays(M, n, r)
+
+        if M.ndim == 0:
+            return rvs1(M, n, r, size, random_state)
+
+        res = np.zeros(size)
+
+        # move shape parameters axes to beginning for easier indexing
```
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:677]: The shape of the output is (shape of RVS for a given set of parameters) + (shape of broadcasted parameter arrays).

I iterate over the parameters with `np.ndenumerate`, which gives an index `i`. I want to use `i`  to index the part of the output array corresponding with one set of shape parameter.

I don't know how to do this when the output is in its original shape, so  I use `moveaxis` to change the shape of the output to (shape of broadcasted parameters) + (shape of RVS size for each set of parameters).

If anyone knows of a simpler way of using the indices `i` without having to move the axes around, I'd love to know about it!

File: scipy/stats/_discrete_distns.py [hunk 2]
```diff
@@ -653,6 +654,36 @@ def _argcheck(self, M, n, r):
...
+            if size is None:
+                return rvs.item()
+            return rvs
+
+        M, n, r = np.broadcast_arrays(M, n, r)
+
+        if M.ndim == 0:
+            return rvs1(M, n, r, size, random_state)
+
+        res = np.zeros(size)
```
Comment (mdhaber) [inline scipy/stats/_discrete_distns.py:675]:
```suggestion
        res = np.empty(size)
```

File: scipy/stats/_discrete_distns.py (hunk 1 above)
Comment (mdhaber) [reply scipy/stats/_discrete_distns.py:677]: Factored this part out as a decorator in gh-13330.

Comment (mdhaber) [reply scipy/stats/_discrete_distns.py:677]: The broadcasting behavior is not quite right here, but I think it's fixed in gh-13330. Marking as draft until we can replace this with that decorator.

**Comment by mdhaber** [comment]
@tirthasheshpatel What do you think of this as a way to speed up RVS for `nhypergeom`?

```python3
from scipy.stats import nhypergeom 
%timeit nhypergeom.rvs(22, 7, 13, size=10000)
```
Before:
```
4.11 s ± 7.84 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
After:
```
1.12 ms ± 1.98 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Identical random variates.

**Comment by tirthasheshpatel** [comment]
> @tirthasheshpatel What do you think of this as a way to speed up RVS for `nhypergeom`?

Thanks, @mdhaber. It's really cool. Simple and fast!

**Review by rgommers — APPROVED** [review]
Nice, thanks @mdhaber! And thanks @tirthasheshpatel for the review and verification.

File: scipy/stats/_discrete_distns.py [hunk 3]
```diff
@@ -6,6 +6,7 @@
 from scipy import special
 from scipy.special import entr, logsumexp, betaln, gammaln as gamln, zeta
 from scipy._lib._util import _lazywhere, rng_integers
+from scipy.interpolate import interp1d
```
Comment (rgommers) [inline scipy/stats/_discrete_distns.py:9]: We already had a dependency on `interpolate` in `_continuous_distns.py`, so this is okay.

**Pull request merged**